nym-network-monitor/src/main.rs

```diff
@@ -0,0 +1,201 @@
+
+    /// Lifetime of each client in seconds
+    #[arg(short = 'T', long, default_value_t = 60)]
+    client_lifetime: u64,
+
+    /// Port to listen on
+    #[arg(long, default_value_t = 8080)]
+    port: u16,
+
+    /// Host to listen on
+    #[arg(long, default_value = "127.0.0.1")]
+    host: String,
+
+    /// Path to the topology file
+    #[arg(short, long, default_value = None)]
+    topology: Option<String>,
+
+    /// Path to the environment file
+    #[arg(short, long, default_value = None)]
+    env: Option<String>,
+
+    #[arg(short, long, default_value_t = 10)]
+    mixnet_timeout: u64,
+
+    #[arg(long, default_value_t = false)]
+    generate_key_pair: bool,
+
+    #[arg(long)]
+    private_key: String,
+}
+
+fn generate_key_pair() -> Result<()> {
+    let mut rng = rand::thread_rng();
+    let keypair = nym_crypto::asymmetric::identity::KeyPair::new(&mut rng);
+
+    let mut public_key_file = File::create("network-monitor-public")?;
+    public_key_file.write_all(keypair.public_key().to_base58_string().as_bytes())?;
+
+    let mut private_key_file = File::create("network-monitor-private")?;
+    private_key_file.write_all(keypair.private_key().to_base58_string().as_bytes())?;
+
+    info!("Generated keypair, public key to 'network-monitor-public', and private key to 'network-monitor-private', public key should be whitelisted with the nym-api");
+
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    nym_bin_common::logging::setup_logging();
+
+    let args = Args::parse();
+
+    setup_env(args.env); // Defaults to mainnet if empty
+
+    let cancel_token = CancellationToken::new();
+    let server_cancel_token = cancel_token.clone();
+    let clients = Arc::new(RwLock::new(VecDeque::with_capacity(args.n_clients)));
+
+    if args.generate_key_pair {
+        generate_key_pair()?;
+        std::process::exit(0);
+    }
+
+    let pk = PrivateKey::from_base58_string(&args.private_key)?;
+    PRIVATE_KEY.set(pk).ok();
+
+    TOPOLOGY
+        .set(if let Some(topology_file) = args.topology {
+            NymTopology::new_from_file(topology_file)?
+        } else {
+            NymTopology::new_from_env().await?
+        })
+        .ok();
+
+    MIXNET_TIMEOUT.set(args.mixnet_timeout).ok();
+
+    let spawn_clients = Arc::clone(&clients);
+    tokio::spawn(make_clients(
+        spawn_clients,
+        args.n_clients,
+        args.client_lifetime,
+        TOPOLOGY.get().expect("Topology not set yet!").clone(),
+    ));
+
+    let _server_handle = tokio::spawn(async move {
+        let socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::from_str(&args.host)?), args.port);
+        let server = HttpServer::new(socket, server_cancel_token);
+        server.run(clients).await
+    });
+
+    info!("Waiting for message (ctrl-c to exit)");
+
+    ctrl_c().await?;
+    info!("Received kill signal, shutting down, submitting final batch of metrics");
+
+    submit_metrics().await?;
+
+    cancel_token.cancel();
+    Ok(())
```

Comment:
Sequentially, what happens here is:
- capture interrupt
- submit metrics
- cancel token -> sends shutdown signal to background task running axum server
- immediately return from main

Returning from `main` here [shuts down the runtime, which cancels all background tasks](https://docs.rs/tokio/latest/tokio/task/#cancellation). This means there is almost no time for cancel signal to be respected by axum server and it may be shutting down forcefully.

If you care about this, you could await on server's `JoinHandle` after giving it a shutdown signal (with a possible time out)

All of this is under assumption you care about shutting it down gracefully. Maybe you don't care about what happens after
```rust
submit_metrics().await?;
```
in which case you can disregard this. 

But then, unless I'm mistaken, signalling via cancellation token is pointless - you can just spawn axum task and forget about it - it will get cancelled anyway when main terminates.

Reply:
All good comments, will address them :)